File: backend/app/models/habit.py
from sqlalchemy import Column, String, DateTime, ForeignKey, Enum, Boolean, Text, Date
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
import uuid
import enum
from ..core.database import Base

class HabitCadence(str, enum.Enum):
    daily = "daily"
    weekly = "weekly"
    custom = "custom"

class HabitLogStatus(str, enum.Enum):
    done = "done"
    skipped = "skipped"

class Habit(Base):
    __tablename__ = "habits"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    name = Column(String, nullable=False)
    cadence = Column(Enum(HabitCadence), nullable=False, default=HabitCadence.daily)
    reminder_time_local = Column(String, nullable=True)  # "HH:MM" format
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    # Relationships
    user = relationship("User", back_populates="habits")
    logs = relationship("HabitLog", back_populates="habit", cascade="all, delete-orphan")

class HabitLog(Base):
    __tablename__ = "habit_logs"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    habit_id = Column(UUID(as_uuid=True), ForeignKey("habits.id"), nullable=False)
    date = Column(Date, nullable=False)
    status = Column(Enum(HabitLogStatus), nullable=False)
    notes = Column(Text, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    # Relationships
    habit = relationship("Habit", back_populates="logs")

File: frontend/src/types/habit.ts
export interface Habit {
  id: string;
  name: string;
  description?: string;
  cadence: 'daily' | 'weekly' | 'custom';
  cadence_config?: {
    days_of_week?: number[]; // 0=Sunday, 1=Monday, etc.
    custom_days?: number;
  };
  reminder_enabled: boolean;
  reminder_time_local?: string; // HH:MM format
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface HabitLog {
  id: string;
  habit_id: string;
  date: string; // YYYY-MM-DD format
  status: 'done' | 'skipped';
  notes?: string;
  created_at: string;
}

export interface CreateHabitRequest {
  name: string;
  description?: string;
  cadence: 'daily' | 'weekly' | 'custom';
  cadence_config?: {
    days_of_week?: number[];
    custom_days?: number;
  };
  reminder_enabled: boolean;
  reminder_time_local?: string;
}

export interface CreateHabitLogRequest {
  date: string;
  status: 'done' | 'skipped';
  notes?: string;
}